toolchain/check/testdata/facet/early_rewrites.carbon

```diff
@@ -13,7 +13,71 @@
 // Tests where `ImplWitnessAccess` on the RHS of a rewrite constraint is used in
 // other constraints within the same facet type in ways that require the RHS to
 // be evaluated to a concrete value before the facet type is resolved (such as
-// by passing it as a parameter to a generic class).
+// by passing it as a parameter to a generic class). This can be done either by
+// eagerly reading rewrite constraints from the same facet type, or by relying
+// on implied constraints which are checked later.
+
+// --- fail_todo_implied_constraint_on_self.carbon
+library "[[@TEST_NAME]]";
+
+interface Z {
+  let Z1:! type;
+  let Z2:! type;
+}
+
+class C(T:! Z where .Z1 = ()) {}
+
+interface J {
+  // Implied: .Z1 == ()
+  let J1:! Z where .Z2 = C(.Self);
+}
+
+fn F(T:! J) {}
+// Satisfies the implied constraint in J.J1.
+fn G(T:! Z where .Z1 = () and .Z2 = C(.Self)) {
+  // CHECK:STDERR: fail_todo_implied_constraint_on_self.carbon:[[@LINE+7]]:3: error: cannot convert type `T` that implements `Z where .(Z.Z1) = () and .(Z.Z2) = C(.Self as Z where .(Z.Z1) = ())` into type implementing `J` [ConversionFailureFacetToFacet]
+  // CHECK:STDERR:   F(T);
+  // CHECK:STDERR:   ^~~~
+  // CHECK:STDERR: fail_todo_implied_constraint_on_self.carbon:[[@LINE-6]]:6: note: initializing generic parameter `T` declared here [InitializingGenericParam]
+  // CHECK:STDERR: fn F(T:! J) {}
+  // CHECK:STDERR:      ^
+  // CHECK:STDERR:
+  F(T);
```

Comment:
Is this testing what you wanted? You're passing a type that impls `X` as an argument for a parameter that's supposed to impl `J` -- should you instead be passing a type that impls `J where .J1 = T`?

Reply:
Ah oops, yeah, that was the goal with the class D/impl D as.. stuff in the other cases, which I failed to apply here, done. I think those are the right shape?